site/app/models/gradeable/AutogradingConfig.php

```diff
@@ -0,0 +1,247 @@
+    protected $total_hidden_non_extra_credit = 0;
+    /** @property @var float Total number of hidden extra-credit points for all test cases */
+    protected $total_hidden_extra_credit = 0;
+
+    public function __construct(Core $core, array $details) {
+        parent::__construct($core);
+
+        // Was there actually a config file to read from
+        if ($details === null || $details === []) {
+            throw new \InvalidArgumentException('Provided details were blank or null');
+        }
+
+        $this->max_submission_size = floatval($details['max_submission_size'] ?? 0);
+        $this->max_submissions = intval($details['max_submissions'] ?? 0);
+        $this->assignment_message = Utils::prepareHtmlString($details['assignment_message'] ?? '');
+
+        $this->required_capabilities = $details['required_capabilities'] ?? 'default';
+        $this->max_possible_grading_time = $details['max_possible_grading_time'] ?? -1;
+
+        if (isset($details['testcases'])) {
+            foreach ($details['testcases'] as $idx => $testcase_details) {
+                $testcase = new AutogradingTestcase($this->core, $testcase_details, $idx);
+
+                // Accumulate the points
+                if ($testcase->isHidden()) {
+                    if ($testcase->isExtraCredit()) {
+                        $this->total_hidden_extra_credit += $testcase->getPoints();
+                    } else {
+                        $this->total_hidden_non_extra_credit += $testcase->getPoints();
+                    }
+                } else {
+                    if ($testcase->isExtraCredit()) {
+                        $this->total_non_hidden_extra_credit += $testcase->getPoints();
+                    } else {
+                        $this->total_non_hidden_non_extra_credit += $testcase->getPoints();
+                    }
+                }
+
+                $this->testcases[$idx] = $testcase;
+            }
+        }
+
+        if (isset($details['early_submission_incentive'])) {
+            $this->early_submission_incentive = true;
+            $this->early_submission_message = Utils::prepareHtmlString($details['early_submission_incentive']['message'] ?? '');
+            $this->early_submission_minimum_days_early = intval($details['early_submission_incentive']['minimum_days_early'] ?? 0);
+            $this->early_submission_minimum_points = intval($details['early_submission_incentive']['minimum_points'] ?? 0);
+            foreach ($details['early_submission_incentive']['test_cases'] ?? [] as $testcase) {
+                $this->early_submission_test_cases[] = $this->testcases[$testcase];
+            }
+        }
+
+        // defaults to 1 if no set
+        $num_parts = count($details['part_names'] ?? [1]);
+
+        // defaults to 0 if not set
+        $num_textboxes = count($details['textboxes'] ?? []);
+
+        // Get all of the part names
+        for ($i = 1; $i <= $num_parts; $i++) {
+            $j = $i - 1;
```

Comment:
What's up with this for-loop

Reply:
Its copy and pasted code from the previous loading method.  I'm not 100% sure what the schema is for the `build_{id}.json` file that this is loaded from and I think its somewhat in flux / needs to be refactored.  I'm not really sure why some parts are 1 indexed and some are 0 indexed

Reply:
Well, it's like that in the original code...